### ISSUE TITLE
using Cuda, Libtorch. register_module(ModuleHolder/pimpl.h) error 

### ISSUE THREAD

**Issue by JohnSidKim** [pytorch/pytorch#64908]
when i run the unet without cuda, it worked well
but i tried to run with cuda, it made an error on following structs.

inc = **register_module("inc", std::make_shared<DoubleConv>(n_channels, st_channels))**

 (in pimpl.h)

>  template <
>       typename Head,
>       typename... Tail,
>       typename = typename std::enable_if<
>           !(torch::detail::is_module_holder_of<Head, ContainedType>::value &&
>             (sizeof...(Tail) == 0))>::type>
>   explicit ModuleHolder(Head&& head, Tail&&... tail)
>       : **impl_(new Contained(
>             std::forward<Head>(head),
>             std::forward<Tail>(tail)...)) {}**

How should i fix it? T.T

> struct UNET : public torch::nn::Module
> {
> UNET(int n_channels = 3, int n_classes = 1, int st_channels = 16, bool bilinear = true):n_channels(n_channels), n_classes(n_classes), st_channels(st_channels), bilinear_(bilinear)
> { 
> // bilinear = st_channels
> // parser.add_argument('--st_channels', type=int, default=16, help='첫 convolution 시 feature map channel')
> **inc = register_module("inc", std::make_shared<DoubleConv>(n_channels, st_channels));**
> down1 = register_module("down1", std::make_shared<Down>(st_channels, st_channels*2));
> down2 = register_module("down2", std::make_shared<Down>(st_channels*2, st_channels*4));
> down3 = register_module("down3", std::make_shared<Down>(st_channels*4, st_channels*8));
> if (bilinear_) factor = 2; else factor = 1;
> down4 = register_module("down4", std::make_shared<Down>(st_channels*8, st_channels*16 / factor));
> up1 = register_module<Up>("up1", std::make_shared<Up>(st_channels*16, st_channels*8 / factor, bilinear));
> up2 = register_module<Up>("up2", std::make_shared<Up>(st_channels*8, st_channels*4 / factor, bilinear));
> up3 = register_module<Up>("up3", std::make_shared<Up>(st_channels*4, st_channels*2 / factor, bilinear));
> up4 = register_module<Up>("up4", std::make_shared<Up>(st_channels*2, st_channels, bilinear));
> outc = register_module<OutConv>("outc", std::make_shared<OutConv>(st_channels, n_classes));
> }
> 
> torch::Tensor forward(torch::Tensor x)
> {
> x1 = inc(x);
> x2 = down1(x1);
> x3 = down2(x2);
> x4 = down3(x3);
> x5 = down4(x4);
> x = up1(x5, x4);
> x = up2(x, x3);
> x = up3(x, x2);
> x = up4(x, x1);
> logits = outc(x);
> return logits;
> }
> 
> torch::nn::ModuleHolder<DoubleConv>	inc{ nullptr };
> torch::nn::ModuleHolder<Down> down1{ nullptr }, down2{ nullptr }, down3{ nullptr }, down4{ nullptr };
> torch::nn::ModuleHolder<Up>	up1{ nullptr }, up2{ nullptr }, up3{ nullptr }, up4{ nullptr };
> torch::nn::ModuleHolder<OutConv> outc{ nullptr };
> 
> torch::Tensor x1,x2,x3,x4,x5;
> torch::Tensor logits;
> private:
> int n_channels;
> int n_classes;
> int st_channels;
> bool bilinear_;
> int factor;
> };

**Comment by jbschlosser** [comment]
Hey @JohnSidKim! I suggest posting on the [PyTorch forums](https://discuss.pytorch.org/) as you may have more success getting a question like this answered there.
